File: src/neetcode/arrays/hashing/Solution36.java
package neetcode.arrays.hashing;

import java.util.HashSet;

/**
 * <a href="https://leetcode.com/problems/valid-sudoku/">Valid Sudoku</a>
 */
class Solution36 {
    public boolean isValidSudoku(char[][] board) {
        HashSet<Character>[] rows = new HashSet[9];
        HashSet<Character>[] cols = new HashSet[9];
        HashSet<Character>[] subs = new HashSet[9];

        for (int n = 0; n < 9; n++) {
            rows[n] = new HashSet<>();
            cols[n] = new HashSet<>();
            subs[n] = new HashSet<>();
        }

        for (int r = 0; r < 9; r++) {
            for (int c = 0; c < 9; c++) {
                int sub = (c / 3 + 1) + 3 * (r / 3) - 1;
                Character ch = board[r][c];
                if (ch.equals('.')) {
                    continue;
                } else if (!rows[r].add(ch) || !cols[c].add(ch) || !subs[sub].add(ch)) {
                    return false;
                }
            }
        }
        return true;
    }
}

File: py/nc/arrays/Solution36.py
from typing import List


class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        rows = [set() for _ in range(9)]
        cols = [set() for _ in range(9)]
        subs = [set() for _ in range(9)]
        for r in range(9):
            for c in range(9):
                t = board[r][c]
                if t == '.':
                    continue
                sub = (c // 3 + 1) + 3 * (r // 3) - 1
                if t in rows[r] or t in cols[c] or t in subs[sub]:
                    return False
                rows[r].add(t)
                cols[c].add(t)
                subs[sub].add(t)
        return True